Introduce chapter about the integration of VueJS in an existing module (#927)
PrestaShop/docs: src/content/1.7/modules/concepts/templating/vuejs/_index.md

{{% children %}}

## 1- Creation of a VueJS application in the module directories

The objective of this part is to make VueJS available in the project. 

We will store the VueJS source code in a dedicated folder that will be present in the module only
during the development part. Running VueJS will be possible by compiling and storing the JS files
in the proper assets folder.

During this tutorial, we use a new folder `_dev` that will be created at the root directory of the module.

Installation of VueJS can be done via [several ways](https://vuejs.org/v2/guide/installation.html),
but we will use the CLI option to get all the stack.

* Installation of VueJS CLI tool (globally)

```
$ npm install -g @vue/cli
```

* Creation of the project

```
$ vue create _dev
```

We chose the preset with Vue 2 & Babel for this tutorial.

After the initialization process with some questions and downloads, a whole standalone application
will be created and available in the `_dev` folder. Running npm commands are defined in the `package.json`
file to run or build it.

```
$ cd _dev
$ npm run build
```

In the next parts, this standalone application will be adapted to run inside a PrestaShop page.

> ## 2- Cleanup

Reviewer: ```suggestion

## 2- Cleanup
```

---

> title: Using VueJS

Reviewer 2: ```suggestion
title: Using Vue.js
```

> menuTitle: VueJS

Reviewer 2: ```suggestion
menuTitle: Vue.js
```

weight: 2
chapter: true
---

# Using VueJS

Depending on the PrestaShop version running on the shop, either on the back or the front-office,

> at least one template renderer is available to display content on the page. With Javascript
> scripts, it is possible to update & animate the page content without having to refresh the page.

Reviewer 2: ```suggestion
at least one template renderer is available to display content on the page. With JavaScript, it is possible to update & animate the page content without having to refresh the page.
```

jQuery is available on PrestaShop to help implementing these interface updates.
For complex interfaces, with state management for instance, other frontend options could be considered.

> [VueJS](https://vuejs.org/) is a Javascript Framework that will be helpful in this case.

Reviewer 2: ```suggestion
[VueJS](https://vuejs.org/) is a JavaScript Framework that will be helpful in this case.
```

> [VueJS](https://vuejs.org/) is a Javascript Framework that will be helpful in this case.

Reviewer 2: ```suggestion
[Vue.js](https://vuejs.org/) is a JavaScript Framework that will be helpful in this case.
```

This chapter covers its integration on a module.

{{% children %}}

> ## 1- Creation of a VueJS application in the module directories

Reviewer 2: ```suggestion
## 1- Creation of a Vue.js application in the module directories
```

> The objective of this part is to make VueJS available in the project. 

Reviewer 2: ```suggestion
The objective of this part is to make Vue.js available in the project. 
```

> We will store the VueJS source code in a dedicated folder that will be present in the module only

Reviewer 2: ```suggestion
We will store the Vue.js source code in a dedicated folder that will be present in the module only
```

> during the development part. Running VueJS will be possible by compiling and storing the JS files

Reviewer 2: ```suggestion
during the development part. Running Vue.js will be possible by compiling and storing the JS files
```

in the proper assets folder.

During this tutorial, we use a new folder `_dev` that will be created at the root directory of the module.

> Installation of VueJS can be done via [several ways](https://vuejs.org/v2/guide/installation.html),

Reviewer 2: ```suggestion
Installation of Vue.js can be done via [several ways](https://vuejs.org/v2/guide/installation.html),
```

but we will use the CLI option to get all the stack.

> * Installation of VueJS CLI tool (globally)

Reviewer 2: ```suggestion
* Installation of Vue.js CLI tool (globally)
```

```
$ npm install -g @vue/cli
```

* Creation of the project

```
$ vue create _dev
```

> We chose the preset with Vue 2 & Babel for this tutorial.

Reviewer 2: ```suggestion
We chose the preset with Vue.js 2 & Babel for this tutorial.
```

Author: For this one I'd like to stick to the content displayed on the screen: They mention precisely 'Vue'

Reviewer 2: @Quetzacoalt91 👍

After the initialization process with some questions and downloads, a whole standalone application
will be created and available in the `_dev` folder. Running npm commands are defined in the `package.json`
file to run or build it.

```
$ cd _dev
$ npm run build
```

> At the moment, we have a VueJS application that can run independantly from the module.

Reviewer 2: ```suggestion
At the moment, we have a Vue.js application that can run independantly from the module.
```

> In the next parts, this standalone application will be adapted to run inside a PrestaShop page.

Reviewer 3: ```suggestion
In the next parts, this standalone application will be adapted to run inside a PrestaShop page.

```

Looks like you're using an empty lien everywhere else before a title

Author: Indeed!

## 2- Cleanup

> Supposing that the VueJS workspace is the `_dev` folder:

Reviewer 2: ```suggestion
Supposing that the Vue.js workspace is the `_dev` folder:
```

* Remove the `_dev/public/` folder.
* In the `_dev/package.json` folder, remove the `serve` script.

## 3- Customization of the application for a PrestaShop context

In [the module structure]({{< ref "/1.7/modules/creation/module-file-structure.md" >}}), JavaScript
files must be stored in `views/js/`. We will ask the compiler to write the distributable files in there.

> Create or edit `_dev/vue.config.js` with:

Reviewer 3: Having JS/assets stuff in the views files looks so weird to me, views are designed to be :
- A view component
- Templating kinds of stuff

Author: That is something that could evolve with the next major version of PrestaShop, same thing for the PHP classes that could end in the `src/` folder instead of `classes/`.

But at this time, I'm sticking to the current structure we require on https://devdocs.prestashop.com/1.7/modules/creation/module-file-structure/. As that is the structure we have from the first versions of PS, that would be strange to store generated assets in another path, don't you think?

Reviewer 3: So, the fact that we did this on Blockwishlist is due to the fact that it's a symfony module? Or it's a miss from us?

I agree with the fact that we should stick to what we defined in the past tbh, but as soon as we can change this, it would be way clearer :)

Author: It seems that several modules started to use their own new folder to store their assets, Faceted search has also (another folder)[https://github.com/PrestaShop/ps_facetedsearch/tree/dev/views/dist] for that matter.

The current validation rules still check the files are in `views/js` & `views/css`, so I'd advise to keep them until there's one announcement about the structure.

Reviewer 3: Ok 👍

Testing the compilation of the files is done with:

```
$ npm run dev
```

After some time, a similar output should appear on the terminal once done:

```
$ npm run dev

[...]

 DONE  Compiled successfully in 2679ms                                                                                                                                19:17:27

  File                            Size                                                                 Gzipped

  ../views/js/chunk-vendors.js    121.79 KiB                                                           44.16 KiB
  ../views/js/app.js              5.41 KiB                                                             1.94 KiB

  Images and other types of assets omitted.

 DONE  Build complete. The ../views directory is ready to be deployed.
 INFO  Check out deployment instructions at https://cli.vuejs.org/guide/deployment.html
```

## 4- Display of the application

The application is now ready to be integrated inside a page of PrestaShop.
There are several steps to complete:

* **Adding a div with the proper ID inside a template**

*modules/[module_name]/views/templates/admin/app.tpl*

```tpl

> <link href="{$pathApp|escape:'htmlall':'UTF-8'}" rel=preload as=script>

Reviewer 2: @Quetzacoalt91 is `escape` still needed for back-office in 1.7?

Author: That's what I thought, I might be not up to date

Author: Escape has been enabled everywhere with https://github.com/PrestaShop/PrestaShop/pull/4176, but hen only enable for the front office with https://github.com/PrestaShop/PrestaShop/commit/f2fc2f791782988cf97de7ba58141f02a9340ab8.

I did not see other changes and `$smarty->escape_html` is still `false` by default. That's why I consider that we still need to escape variables on the back office.

Reviewer 2: @Quetzacoalt91 I think you're right 👍

<div id="app"></div>
<script src="{$chunkVendor|escape:'htmlall':'UTF-8'}"></script>
<script src="{$pathApp|escape:'htmlall':'UTF-8'}"></script>
```

Although this is not recommended, this example adds the JS file inside the template so the example remains short.

> Prefer using `registerJavascript()` when possible.

Reviewer 2: ```suggestion
You should use `registerJavascript()` when possible.
```

* **Displaying the template content**

For an application integrated in an `AdminModuleController`:
```php
public function initContent()
{
  $this->context->smarty->assign([
      'pathApp' => $this->module->getPathUri() . 'views/js/app.js',
      'chunkVendor' => $this->module->getPathUri() . 'views/js/chunk-vendors.js',
  ]);

>   $this->content = $this->context->smarty->fetch($this->module->getLocalPath() . '/views/templates/admin/app.tpl');

Reviewer 2: ```suggestion
  $this->content = $this->context->smarty->fetch('module:modulename/views/templates/admin/app.tpl');
```

  parent::initContent();
}
```

Or for an application integrated in the module main class (i.e in the method `getContent()`):

*modules/[module_name]/[module_name].php*

```php
public function getContent()
{
  $this->context->smarty->assign([
      'pathApp' => $this->getPathUri() . 'views/js/app.js',
      'chunkVendor' => $this->getPathUri() . 'views/js/chunk-vendors.js',
  ]);

>   return $this->context->smarty->fetch($this->getLocalPath() . 'views/templates/admin/app.tpl');

Reviewer 2: ```suggestion
  return $this->context->smarty->fetch('module:modulename/views/templates/admin/app.tpl');
```

}
```

> If the integration goes fine, the page should display the welcome screen of VueJS:

Reviewer 2: ```suggestion
If the integration goes fine, the page should display the welcome screen of Vue.js:
```

> {{< figure src="../img/bo-with-vuejs.png" title="Module configuration page with welcome page of VueJS" >}}

Reviewer 2: ```suggestion
{{< figure src="../img/bo-with-vuejs.png" title="Module configuration page with welcome page of Vue.js" >}}
```

The application is now running and ready to be modified with the module features.

> The next steps can be done by relying on the [VueJS documentation](https://vuejs.org/v2/guide/) or website

Reviewer 2: ```suggestion
The next steps can be done by relying on the [Vue.js documentation](https://vuejs.org/v2/guide/) or website
```
